### PR TITLE
refactor: wrap up rewrite in TS

### PR DESCRIPTION
**Pull request by vladfrangu**

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -16,3 +19,36 @@ export type UnwrapPromise<T> = T extends PromiseLike<infer R> ? UnwrapPromise<R>
 
 // eslint-disable-next-line @typescript-eslint/no-empty-function, @typescript-eslint/no-unused-vars
 export function noop(..._args: unknown[]): void {}
+
+export type InferBrowserPluginArray<
```
Comment (pocesar) [inline src/utils.ts:23]: this is considered art 🤩 
what is the current impact on the language server when unwrapping the array?

Comment (vladfrangu) [reply src/utils.ts:23]: That... I don't know, I've never actually measured that before 😓 
Any tips on how to do it?

File: src/abstract-classes/browser-plugin.ts [hunk 2]
```diff
@@ -25,7 +25,7 @@ export interface CommonLibrary {
  * @internal
  */
 export interface CommonBrowser {
-    newPage(...args: unknown[]): unknown;
+    newPage(...args: unknown[]): any;
```
Comment (szmarczak) [inline src/abstract-classes/browser-plugin.ts:28]: Why the change?

Comment (vladfrangu) [reply src/abstract-classes/browser-plugin.ts:28]: Tried making an interface for it too, but I'd always get errors from WeakMap, and it doesn't really matter since this is overwritten by generics everywhere
image

Comment (szmarczak) [reply src/abstract-classes/browser-plugin.ts]: It do be like that sometimes. Glad this is now resolved :)

File: src/utils.ts (hunk 1 above)
Comment (szmarczak) [reply src/utils.ts:23]: That shouldn't have a significant impact imo. Many other packages do unwrapping as well.

File: src/abstract-classes/browser-plugin.ts [hunk 3]
```diff
@@ -21,11 +21,14 @@ export interface CommonLibrary {
...
- * @internal
- */
+/** @internal */
 export interface CommonBrowser {
-    newPage(...args: unknown[]): unknown;
+    newPage(...args: unknown[]): Promise<CommonPage>;
+}
+
+/** @internal */
+export interface CommonPage {
```
Comment (szmarczak) [inline src/abstract-classes/browser-plugin.ts:30]: Can't this be? I'm don't know much about the internals yet so just asking :)

```ts
import { Page as PuppeteerPage } from 'puppeteer';
import { Page as PlaywrightPage } from 'playwright';

export type CommonPage = PuppeteerPage | PlaywrightPage;
```

Comment (vladfrangu) [reply src/abstract-classes/browser-plugin.ts:30]: Could it, yes, but since either dependencies can be missing at runtime, I think it'd widen to `any` which is... not helpful at all

File: src/browser-pool.ts [hunk 4]
```diff
@@ -487,7 +570,7 @@ class BrowserPool extends EventEmitter {
             // If the hooks or the launch fails, we need to delete the controller,
             // because otherwise it would be stuck in limbo without a browser.
             await this._executeHooks(this.preLaunchHooks, pageId, launchContext);
-            const browser = await browserPlugin.launch(launchContext);
+            const browser = await browserPlugin.launch(launchContext as never);
```
[the remark this answers is not in the document]
Comment (vladfrangu) [reply src/browser-pool.ts:573]: Nice catch! However, it's still needed in tests

File: src/abstract-classes/browser-plugin.ts (hunk 3 above)
Comment (szmarczak) [reply src/abstract-classes/browser-plugin.ts:30]: Woah, you're right.

Comment (szmarczak) [reply src/abstract-classes/browser-plugin.ts:30]: It's a pity they don't export just the types as a separate package... That's the con of big projects written in TS.

**Review by szmarczak — APPROVED** [review]
LGTM
